lib/components/table.vue

```diff
@@ -283,27 +370,82 @@
+            rowClicked(e, item, index) {
+                if (this.busy) {
+                    // If table is busy (via provider) then don't propagate
+                    e.preventDefault();
+                    e.stopPropagation();
+                    return;
+                }
                 this.$emit('row-clicked', item, index);
             },
-            headClick(field, key) {
+            headClick(e, field, key) {
+                if (this.busy) {
+                    // If table is busy (via provider) then don't propagate
+                    e.preventDefault();
+                    e.stopPropagation();
+                    return;
+                }
                 if (!field.sortable) {
                     this.sortBy = null;
+                } else {
+                    if (key === this.sortBy) {
+                        this.sortDesc = !this.sortDesc;
+                    } else {
+                        this.sortDesc = true;
+                    }
+                    this.sortBy = key;
+                }
+                this.$emit('head-clicked', key, this.sortDesc);
+            },
+            refresh() {
+                // Expose refresh method
+                if (this.hasProvider) {
+                    this._updateProvider()
+                }
+            },
+            _providerSetLocal(items) {
+                this.localItems = (items && items.length > 0) ? items.slice() : [];
+                this.$emit('refreshed');
+                this.$root.$emit('table::refreshed', this.id);
+            },
+            _providerUpdate() {
+                // Refresh the provider items
+                if (this.busy || !this.hasProvider) {
+                    // Don't refresh remote data if we are 'busy' or if no provider
                     return;
                 }
 
-                if (key === this.sortBy) {
-                    this.sortDesc = !this.sortDesc;
+                // Call provider function with context
+                const data = this.items({
+                    perPage: this.perPage,
+                    currentPage: this.currentPage,
+                    filter: this.filter,
+                    sortBy: this.sortBy,
+                    sortDesc: this.sortDesc,
+                    callback: this._providerSetLocal
+                });
+
+                if (!data) {
+                    // Provider is using callback
+                    return;
                 }
 
-                this.sortBy = key;
+                if (data.then && typeof data.then === 'function') {
+                    // Provider returned Promise
+                    data.then((items) => {
+                        this._providerSetLocal(items);
+                    });
+                } else {
+                    // Provider returned Array data
+                    this._providerSetLocal(data);
+                }
             }
         }
     };
 </script>
 
-
-<style>
+<style scoped>
```

Comment:
Scopes are inconsistent in  bundlers. We may simply prefix root div and selectors with something like `.bs-table` class. (Or maybe using [css-modules](https://vue-loader.vuejs.org/en/features/css-modules.html))